rust/operator-binary/src/controller/build.rs

```diff
@@ -0,0 +1,470 @@
+mod config;
+mod discovery;
+mod extensions;
+mod internal_secret;
+mod roles;
+
+use discovery::build_discovery_configmaps;
+use internal_secret::build_shared_internal_secret;
+
+use snafu::{ResultExt, Snafu};
+use stackable_druid_crd::{
+    ldap::DruidLdapSettings, resource, security::DruidTlsSecurity, DruidCluster, DruidRole,
+};
+use stackable_operator::{
+    kube::runtime::reflector::ObjectRef,
+    product_config::ProductConfigManager,
+    product_config_utils::{transform_all_roles_to_config, validate_all_roles_and_groups_config},
+    role_utils::RoleGroupRef,
+};
+use std::{str::FromStr, sync::Arc};
+use strum::{EnumDiscriminants, IntoStaticStr};
+
+use self::roles::{
+    build_role_service, build_rolegroup_config_map, build_rolegroup_services,
+    build_rolegroup_statefulset,
+};
+
+use super::{
+    types::{AdditionalData, AppliableClusterResource},
+    CONTROLLER_NAME,
+};
+
+#[derive(Snafu, Debug, EnumDiscriminants)]
+#[strum_discriminants(derive(IntoStaticStr))]
+#[allow(clippy::enum_variant_names)]
+pub enum Error {
+    #[snafu(display("invalid product config"))]
+    InvalidProductConfig {
+        source: stackable_operator::error::Error,
+    },
+    #[snafu(display("failed to transform configs"))]
+    ProductConfigTransform {
+        source: stackable_operator::product_config_utils::ConfigError,
+    },
+    #[snafu(display("failed to build discovery ConfigMap"))]
+    BuildDiscoveryConfig { source: discovery::Error },
+    #[snafu(display("could not parse Druid role [{role}]"))]
+    UnidentifiedDruidRole {
+        source: strum::ParseError,
+        role: String,
+    },
+    #[snafu(display("failed to resolve and merge resource config for role and role group"))]
+    FailedToResolveResourceConfig {
+        source: stackable_druid_crd::resource::Error,
+    },
+    #[snafu(display("failed to retrieve secret for internal communications"))]
+    FailedInternalSecretCreation { source: internal_secret::Error },
+    #[snafu(display("failed to build role related cluster resource"))]
+    RoleBuildError { source: roles::Error },
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub async fn create_appliable_cluster_resources(
```

Comment:
As far as I see, this function doesn't need to be `async` (takes no `client`)

Reply:
Good point! I took this for granted. Couldn't resist fixing this.